crates/noirc_evaluator/src/lib.rs

```diff
@@ -40,33 +40,44 @@ pub fn create_circuit(
     is_blackbox_supported: IsBlackBoxSupported,
     enable_logging: bool,
     show_output: bool,
-) -> Result<Circuit, RuntimeError> {
+) -> Result<(Circuit, Abi), RuntimeError> {
     let mut evaluator = Evaluator::new();
 
     // First evaluate the main function
-    evaluator.evaluate_main_alt(program, enable_logging, show_output)?;
+    evaluator.evaluate_main_alt(program.clone(), enable_logging, show_output)?;
 
     let witness_index = evaluator.current_witness_index();
 
+    let mut abi = program.abi;
+    abi.param_witnesses = evaluator.param_witnesses;
+
+    let public_inputs = abi
+        .clone()
+        .public_abi()
+        .parameter_names()
+        .into_iter()
+        .flat_map(|param_name| abi.param_witnesses[param_name].clone())
+        .collect();
+
     let optimized_circuit = acvm::compiler::compile(
         Circuit {
             current_witness_index: witness_index,
             opcodes: evaluator.opcodes,
-            public_inputs: PublicInputs(evaluator.public_inputs),
+            public_inputs: PublicInputs(public_inputs),
         },
         np_language,
         is_blackbox_supported,
     )
     .map_err(|_| RuntimeErrorKind::Spanless(String::from("produced an acvm compile error")))?;
 
-    Ok(optimized_circuit)
+    Ok((optimized_circuit, abi))
 }
 
 impl Evaluator {
     fn new() -> Self {
         Evaluator {
-            public_inputs: Vec::new(),
             num_witnesses_abi_len: 0,
+            param_witnesses: BTreeMap::new(),
```

Comment:
This makes sense, since the only way to get a public input is by through main parameters and return